### ISSUE TITLE
Due USB device crashes after ~72 min

### ISSUE THREAD

**Issue by alterscape** [jmriego/Fino#24]
I'm building a custom Due-based sketch derived from Fino (thanks again for Due support)! I noticed that the USB device dies after about 72 minutes, which is very suspicious of a 32-bit unsigned int tracking micros overflowing, somewhere (don't know where, can't find any likely candidates on my own).  I just finally got around to testing with just you original Fino code, w/o my additions, and I still encounter the same crash. I've tested on two different Due boards so I suspect it's the code, not the hardware. Any ideas where one might look? (Unfortunately it's a real pain to test..)

Additional context: the sketch keeps running and serial/etc code continues to be executed as expected, but Windows reports that the USB device has disconnected, and then when it reconnects, that it is not a valid USB device. Connecting and disconnecting the USB cable results in the same error, until the microcontroller is reset, either by fully removing power or tapping the reset button.

**Comment by jmriego** [comment]
oh wow, that's really strange. As you said, that totally looks like a number of micros overflow in an unsigned long variable.
The maximum value of that would be 2^32. That would give you around 72 minutes if the variables were in micros, but 52 days if using millis, which is actually what I'm using. A quick search in my code hasn't returned any variable with micros, just millis.

Anyway, I'll try to do some testing to see if I can figure something out here.

Can you explain a bit more what you mean by "Connecting and disconnecting the USB cable results in the same error, until the microcontroller is reset, either by fully removing power or tapping the reset button."? The only thing I'm not clear is what you mean by "fully removing power". Is that not the same as disconnecting the USB cable or are you using another power source? Also, I guess you are using Windows, right? Just in case I might test in Linux too

**Comment by alterscape** [comment]
Apologies for being unclear! I'm using both USB ports on the Due (the "programming" port for serial, and the other port for the joystick).  What I mean is that, if I leave the programming USB cable connected so the device is still getting power, I can disconnect and reconnect the joystick USB, while still powering the board with the programming port.  When I do that, Windows reports each reconnect as "something is wrong with the USB device you just connected."  If I disconnect both USB cables, of course the device resets.

So far this is all Windows, and all the same machine. I can test with a spare laptop to make sure  it's not my gaming/dev PC, and I'm pretty sure I have or can create an Ubuntu USB stick to test that too. I also need to check the Event Log to see what Windows is calling the error.

The other thing I could try, I suppose, is to make the same modifications you did to the original (non-FFB) pluggableUSB joystick library and see if the crash reproduces there, which could help isolate the problem...

**Comment by alterscape** [comment]
I suspect this might be some edge-case in the Due HAL,  since I agree that the Fino code isn't doing anything with micros, except for a single delayMicroseconds() call in DynamicUSB, which, as far as I can tell, is only ever invoked on setup, not repeatedly in the loop!

**Comment by jmriego** [comment]
No worries! We'll end up figuring this one out.
I have been running my Arduino Due for 4 hours now and it's still showing up in Windows
![image](https://user-images.githubusercontent.com/15045950/173238394-486edb32-37a8-45e8-9444-b251b13ce3ee.png)

So the USB device disappears for you, right?

I have run it with the code just as it is in the repo right now. I understand that the code as it is won't be too useful for you as you'd need to attach your sensors, etc, but can you test anyway if this repo works without any of your changes? If that works, at least we're getting close and I'd be happy to check your code for anything that I can suspect causes this

**Comment by alterscape** [comment]
Ok _that_ is interesting! Yesterday I tried building from your `master` branch (without any of my changes), and I encountered the same error there. That points to an issue with my PC, rather than the Due or the firmware! I will test again on another PC today, again using your unmodified firmware.

**Comment by alterscape** [comment]
Update: I've had my second Due running Fino `master` plugged in to my laptop (also Win10) for about 80 minutes, which is long enough to trigger the fault I've encountered.  So, it seems to be something specific to my PC, and not a Fino issue at all!

Next troubleshooting steps I'm going to try: connect to a USB2 port, just in case it's weirdness with the usb3 controller/drivers on my sim/gaming PC (it's an Aorus Z390 Pro Wifi mobo; using the onboard Intel USB3.1 controller so far). 

If that fails, I can also try booting to Ubuntu from a USB stick, to see if maybe it's a Windows/driver thing. If it doesn't work, that doesn't tell me too much (could be hardware problems or the same bug in both Linux and windows drivers), but if it works, then it suggests it's a driver issue.

If it works on Ubuntu and not Windows, then I guess it's time to try Windows 11 and/or a new mainboard, hah.

**Comment by alterscape** [comment]
Hm, the problem still occurs on my Windows machine, even using a different USB controller:

```
Windows has stopped this device because it has reported problems. (Code 43)

A request for the USB device descriptor failed.
```

**Comment by jmriego** [comment]
I really can't think of any reason for this to happen. I've been googling this but nothing similar comes up.
can you try connecting only the USB cable next to the reset button? Also, it might be worth it testing with the newest Arduino IDE (v2) or maybe checking if you get this error when you flash another completely different sketch to the Due

**Comment by jmriego** [comment]
Maybe you can also try completely deleting the drivers. You could try using USBDeview and going to the Arduino Due (I think it should have vendor id 2341). Then right clock on all devices you see for it (it should find the Due and the joystick at least), and uninstall selected devices

**Comment by alterscape** [comment]
Bonus  fun: I tried assigning the Due HID device to an Ubuntu instance running in VirtualBox, and.. works fine there too!

I've tried the uninstall mechanism you suggest with USBDeview, I'll see if that works!

**Comment by alterscape** [comment]
USBDeview delete seems to have resolved the issue, at least for the build of `master` on the second Due. I'll try connecting my controller Due and USBDeview deleting there too, and close if it works! Thank you for the suggestion!

**Comment by jmriego** [comment]
that was close! I find it so strange that you even got it to work inside a virtual machine. Honestly I can't think of many things you could check.
It's also possible to change the VID and PID of your Arduino Due so that Windows would see it as a totally new device. Not sure if that would fix it as you have deleted the drivers and didn't seem to work though. But anyway, just in case, this link explains it. Happy to get into more detail if you get stuck http://liveelectronics.musinou.net/MIDIdeviceName.php

**Comment by alterscape** [comment]
I've identified what differs between my laptop (where this works flawlessly) and my desktop (where I observe this fault):  Under the Device Manger Property Page, "Events," I see the following error. Note that I used the instructions you referenced to change the PID/VID, so the VID and PID are custom (0x1209, 0x0001), but I observe the same error.  This error does not occur on my laptop, where the device works fine indefinitely.

```
Device HID\VID_1209&PID_0001&MI_02\8&13464a7c&0&0000 was not migrated due to partial or ambiguous match.

Last Device Instance Id: USB\VID_24F0&PID_0137&MI_00\7&1a0e9595&0&0000
Class Guid: {745a17a0-74d3-11d0-b6fe-00a0c90f57da}
Location Path: 
Migration Rank: 0xF000FFFFFFFFF120
Present: false
Status: 0xC0000719
```

I've seen some references that using a USB 2.0 hub sometimes resolves this issue for Adafruit SAM-based boards, so I'm trying that. If that doesn't work I think next step is an OS reinstall.
